pkg/controller/volume/attachdetach/populator/desired_state_of_world_populator.go

```diff
@@ -154,6 +154,22 @@ func (dswp *desiredStateOfWorldPopulator) findAndRemoveDeletedPods() {
 		klog.V(1).Infof("Removing pod %q (UID %q) from dsw because it does not exist in pod informer.", dswPodKey, dswPodUID)
 		dswp.desiredStateOfWorld.DeletePod(dswPodUID, dswPodToAdd.VolumeName, dswPodToAdd.NodeName)
 	}
+
+	// check if the existing volumes changes its attachability
+	for _, volumeToAttach := range dswp.desiredStateOfWorld.GetVolumesToAttach() {
+		// IsAttachableVolume() will result in a fetch of CSIDriver object if the volume plugin type is CSI.
+		// The result is returned from CSIDriverLister which is from local cache. So this is not an expensive call.
+		volumeAttachable := volutil.IsAttachableVolume(volumeToAttach.VolumeSpec, dswp.volumePluginMgr)
+		if !volumeAttachable {
+			klog.Infof("Volume %v changes from attachable to non-attachable.", volumeToAttach.VolumeName)
+			for _, scheduledPod := range volumeToAttach.ScheduledPods {
+				podUID := volutil.GetUniquePodName(scheduledPod)
+				dswp.desiredStateOfWorld.DeletePod(podUID, volumeToAttach.VolumeName, volumeToAttach.NodeName)
+				klog.V(4).Infof("Removing podUID: %v, volume: %v on node: %v from desired state of world"+
+					" because of the change of volume attachability.", podUID, volumeToAttach.VolumeName, volumeToAttach.NodeName)
+			}
+		}
+	}
```

Comment:
we already have node, pod and volume name from `GetPodToAdd` function. Shouldn't we able to get volumeSpec from that and integrate removal of pod from DSOW in the loop above rather than creating a new loop?

Reply:
IMHO, the way `GetPodToAdd` returns the pods list but that does not guarantee to have all the volumes we are referring to. There might be cases where pod A mounted volume x, y. And in the `GetPodToAdd` function, the final return pod list will only have a map of pod A to volume y. Because volume y has been overridden. Secondly, there might be cases where volume x has two corresponding pods. If we check the volume from the `GetPodToAdd` loop. There will be an overhead of `IsAttachableVolume` check for a volume.

Does this makes sense?

Reply:
hmm, the first issue sounds like inefficiency in populator. if a pod has more than 1 volume - it may require more than one reconciliation loop for both volumes and pod to disappear from DSOW, which could be 1 minute apart.  IMO - it is worth fixing that, but we can file this as a separate issue.